### DISCUSSION TITLE
return cosine score for FaissSearcher

### DISCUSSION THREAD

**Discussion by dayuyang1999** [castorini/pyserini#1951] [Q&A] [answered]
Hi,

I use default FaissSearcher.search, the hits.score is a score from an arbitrary range. I wish to only return me a score within range [-1, 1] (as cosine similarity score).

Do you know how can I achieve this?

**Comment by lintool** [comment]
Most models are trained with dot product similarity, so you'd have to do normalization yourself. Standard approach of normalizing wrt to observed min and max values works fine in most cases. Try that first?

**Reply by dayuyang1999** [reply]
Hi,

Thank you so much for the information.

I am wondering why the standard approach is min-max normalization instead of dividing the norm of two vectors and converting the score to cosine similarity?

**Reply by lintool** [reply]
You can't do that if the model was trained on dot product similarity.

**Reply by dayuyang1999** [reply]
Thank you, yes I agree model trained on dot product cannot produce cosine similarity because after dividing the L2 norm of each embedding, the similarity value may changed.

But I am wondering:
1. Why people chose dot product to train instead of cosine similarity? 
- dot product considers the magnitude of embeddings...but what are the pros?
2. Do most of the embedding models trained on dot product? or cosine similarity?
- I was using bge, gte, e5-mistral,... to compute the cosine similarity between documents. Seems I cannot do that if they are trained on dot product.

**Reply by lintool** [reply]
**Empirically**, dot product vs. cosine, I don't think it really matters, just an arbitrary choice. @MXueguang you have more to chime in here?

Some of the recent models _are_ trained with cosine similarity - you'll have to read the papers for details.

**Reply by dayuyang1999** [reply]
Hi,

Thank you so much for the information!

I checked e5-mistral-7b-instruct paper, they claimed they use cosine similarity

<img width="553" alt="image" src="https://github.com/user-attachments/assets/ba9d97e2-fad3-4bf0-92fb-2fe3461306a9">


Does that mean when I create index using pyserini, I must use `--l2-norm` argument?

**Reply by lintool** [reply, marked as answer]
I believe so.
